File: backend/app/routers/listings.py
import uuid
from decimal import Decimal

from fastapi import APIRouter, Depends, Query, Request, UploadFile, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.dependencies import (
    get_current_admin,
    get_current_user,
    get_optional_user,
    get_seller,
)
from app.core.rate_limit import client_identifier
from app.db.session import get_db
from app.models.listing import Condition
from app.models.shop import Shop
from app.models.user import User
from app.schemas.common import Page
from app.schemas.listing import (
    ListingCreate,
    ListingFeatureIn,
    ListingImageOut,
    ListingOut,
    ListingUpdate,
)
from app.schemas.rating import RatingOut, SellerReviewsOut
from app.services import listing_service, media_service, rating_service

router = APIRouter(prefix="/listings", tags=["listings"])


@router.post("", response_model=ListingOut, status_code=status.HTTP_201_CREATED)
async def create_listing(
    payload: ListingCreate, user: User = Depends(get_seller), db: AsyncSession = Depends(get_db)
) -> ListingOut:
    listing = await listing_service.create_listing(db, user, payload)
    return ListingOut.model_validate(listing)


@router.get("", response_model=Page[ListingOut])
async def browse_listings(
    q: str | None = None,
    tags: list[str] = Query(default=[]),
    min_price: Decimal | None = None,
    max_price: Decimal | None = None,
    condition: Condition | None = None,
    shop_id: uuid.UUID | None = None,
    seller_id: uuid.UUID | None = None,
    personal_only: bool = False,
    is_top: bool | None = None,
    category_id: uuid.UUID | None = None,
    category: str | None = Query(default=None, description="Category slug; includes subcategories"),
    sort: str = "newest",
    limit: int = 20,
    offset: int = 0,
    db: AsyncSession = Depends(get_db),
) -> Page[ListingOut]:
    filters = listing_service.BrowseFilters(
        q=q,
        tags=tags,
        min_price=min_price,
        max_price=max_price,
        condition=condition,
        shop_id=shop_id,
        seller_id=seller_id,
        personal_only=personal_only,
        is_top=is_top,
        category_id=category_id,
        category_slug=category,
        sort=sort,
        limit=limit,
        offset=offset,
    )
    items, total = await listing_service.browse_listings(db, filters)
    return Page(
        items=[ListingOut.model_validate(i) for i in items], total=total, limit=filters.limit, offset=offset
    )


@router.get("/suggestions", response_model=list[str])
async def get_search_suggestions(
    q: str = Query(..., min_length=2),
    limit: int = Query(default=5, le=10),
    db: AsyncSession = Depends(get_db),
) -> list[str]:
    """Returns a list of search suggestions for the given query."""
    return await listing_service.get_search_suggestions(db, q, limit)


@router.get("/mine", response_model=list[ListingOut])
async def list_my_listings(
    shop_id: uuid.UUID | None = None,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[ListingOut]:
    listings = await listing_service.list_my_listings(db, user.id, shop_id)
    return [ListingOut.model_validate(i) for i in listings]


@router.get("/{listing_id}", response_model=ListingOut)
async def get_listing(
    listing_id: uuid.UUID,
    request: Request,
    viewer: User | None = Depends(get_optional_user),
    db: AsyncSession = Depends(get_db),
) -> ListingOut:
    listing = await listing_service.get_listing(db, listing_id)
    # Only the detail route counts. /related and /seller-reviews also load a
    # listing, and counting those would inflate the number with traffic the
    # seller never actually received.
    await listing_service.record_view(
        db,
        listing,
        viewer,
        client_identifier(request),
        request.headers.get("user-agent"),
    )
    return ListingOut.model_validate(listing)


@router.get("/{listing_id}/related", response_model=list[ListingOut])
async def get_related_listings(
    listing_id: uuid.UUID,
    limit: int = Query(default=8, le=24),
    db: AsyncSession = Depends(get_db),
) -> list[ListingOut]:
    listing = await listing_service.get_listing(db, listing_id)
    related = await listing_service.list_related_listings(db, listing, limit)
    return [ListingOut.model_validate(i) for i in related]


@router.get("/{listing_id}/seller-reviews", response_model=SellerReviewsOut)
async def get_seller_reviews(listing_id: uuid.UUID, db: AsyncSession = Depends(get_db)) -> SellerReviewsOut:
    """Reviews for whoever is selling this listing — the shop if it's a shop
    listing, otherwise the individual seller. Same split the rating system
    itself uses when deciding what a rating targets."""
    listing = await listing_service.get_listing(db, listing_id)

    if listing.shop_id is not None:
        shop = await db.get(Shop, listing.shop_id)
        average, count = await rating_service.get_shop_rating_summary(db, shop.id)
        breakdown = await rating_service.get_star_breakdown(db, shop_id=shop.id)
        reviews = await rating_service.list_shop_ratings(db, shop.id)
        target_type, target_name, target_url = "shop", shop.shop_name, f"/shops/{shop.slug}"
    else:
        seller = listing.seller
        average, count = await rating_service.get_user_rating_summary(db, seller.id)
        breakdown = await rating_service.get_star_breakdown(db, user_id=seller.id)
        reviews = await rating_service.list_user_ratings(db, seller.id)
        target_type, target_name, target_url = "user", seller.full_name, f"/profile/{seller.id}"

    return SellerReviewsOut(
        target_type=target_type,
        target_name=target_name,
        target_url=target_url,
        average_rating=average,
        rating_count=count,
        breakdown=breakdown,
        reviews=[RatingOut.model_validate(r) for r in reviews],
    )


@router.patch("/{listing_id}", response_model=ListingOut)
async def update_listing(
    listing_id: uuid.UUID,
    payload: ListingUpdate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ListingOut:
    listing = await listing_service.get_owned_listing(db, listing_id, user)
    listing = await listing_service.update_listing(db, listing, payload)
    return ListingOut.model_validate(listing)


@router.delete("/{listing_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_listing(
    listing_id: uuid.UUID, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)
) -> None:
    listing = await listing_service.get_owned_listing(db, listing_id, user)
    await listing_service.delete_listing(db, listing)


@router.post("/{listing_id}/mark-sold", response_model=ListingOut)
async def mark_sold(
    listing_id: uuid.UUID, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)
) -> ListingOut:
    listing = await listing_service.get_owned_listing(db, listing_id, user)
    listing = await listing_service.mark_sold(db, listing)
    return ListingOut.model_validate(listing)


@router.post("/{listing_id}/renew", response_model=ListingOut)
async def renew_listing(
    listing_id: uuid.UUID, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)
) -> ListingOut:
    # get_owned_listing, not get_listing: an expired listing is still is_active
    # and not soft-deleted, so its owner can reach and revive it.
    listing = await listing_service.get_owned_listing(db, listing_id, user)
    listing = await listing_service.renew_listing(db, listing)
    return ListingOut.model_validate(listing)


@router.post("/{listing_id}/feature", response_model=ListingOut)
async def feature_listing(
    listing_id: uuid.UUID,
    payload: ListingFeatureIn,
    _admin: User = Depends(get_current_admin),
    db: AsyncSession = Depends(get_db),
) -> ListingOut:
    """Admin-only. Promotion is granted, not bought — there's no payment
    integration, so letting sellers set this themselves would just mean
    everyone is featured and the ordering means nothing."""
    listing = await listing_service.get_listing(db, listing_id)
    listing = await listing_service.set_featured(db, listing, payload.days)
    return ListingOut.model_validate(listing)


@router.post("/{listing_id}/images", response_model=ListingImageOut, status_code=status.HTTP_201_CREATED)
async def upload_listing_image(
    listing_id: uuid.UUID,
    file: UploadFile,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ListingImageOut:
    listing = await listing_service.get_owned_listing(db, listing_id, user)
    image_url = await media_service.save_image(file, "listings")
    image = await listing_service.add_image(db, listing, image_url)
    return ListingImageOut.model_validate(image)


@router.delete("/{listing_id}/images/{image_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_listing_image(
    listing_id: uuid.UUID,
    image_id: uuid.UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> None:
    listing = await listing_service.get_owned_listing(db, listing_id, user)
    await listing_service.delete_image(db, listing, image_id)

File: frontend/lib/api/listings.ts
import { apiFetch } from "@/lib/api/client";
import type {
  Condition,
  FulfillmentType,
  Listing,
  ListingImage,
  Page,
  PriceType,
  SellerReviews,
} from "@/types/api";

export interface ListingPayload {
  title: string;
  description: string;
  price?: number | null;
  price_type: PriceType;
  unit?: string | null;
  condition?: Condition | null;
  shop_id?: string | null;
  category_id?: string | null;
  tags: string[];
  fulfillment_type: FulfillmentType;
  pickup_address?: string | null;
}

export function createListing(payload: ListingPayload) {
  return apiFetch<Listing>("/listings", { method: "POST", body: JSON.stringify(payload) });
}

export interface BrowseFilters {
  q?: string;
  tags?: string[];
  min_price?: number;
  max_price?: number;
  condition?: Condition;
  shop_id?: string;
  seller_id?: string;
  personal_only?: boolean;
  is_top?: boolean;
  category_id?: string;
  category?: string;
  sort?: "newest" | "price_asc" | "price_desc" | "popular" | "manual";
  limit?: number;
  offset?: number;
}

export function browseListings(filters: BrowseFilters = {}) {
  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(filters)) {
    if (value === undefined || value === "") continue;
    if (Array.isArray(value)) {
      value.forEach((v) => params.append(key, v));
    } else {
      params.append(key, String(value));
    }
  }
  const qs = params.toString();
  return apiFetch<Page<Listing>>(`/listings${qs ? `?${qs}` : ""}`);
}

export function getSearchSuggestions(q: string, limit = 5) {
  return apiFetch<string[]>(`/listings/suggestions?q=${encodeURIComponent(q)}&limit=${limit}`);
}


export function getMyListings(shopId?: string | null) {
  const qs = shopId ? `?shop_id=${shopId}` : "";
  return apiFetch<Listing[]>(`/listings/mine${qs}`);
}

export function getListing(id: string) {
  return apiFetch<Listing>(`/listings/${id}`);
}

export function getRelatedListings(id: string, limit = 8) {
  return apiFetch<Listing[]>(`/listings/${id}/related?limit=${limit}`);
}

export function getSellerReviews(id: string) {
  return apiFetch<SellerReviews>(`/listings/${id}/seller-reviews`);
}

export function updateListing(id: string, payload: Partial<ListingPayload>) {
  return apiFetch<Listing>(`/listings/${id}`, { method: "PATCH", body: JSON.stringify(payload) });
}

export function deleteListing(id: string) {
  return apiFetch<void>(`/listings/${id}`, { method: "DELETE" });
}

export function markSold(id: string) {
  return apiFetch<Listing>(`/listings/${id}/mark-sold`, { method: "POST" });
}

export function renewListing(id: string) {
  return apiFetch<Listing>(`/listings/${id}/renew`, { method: "POST" });
}

export function markOutOfStock(id: string) {
  return apiFetch<Listing>(`/listings/${id}/mark-out-of-stock`, { method: "POST" });
}

export function markAvailable(id: string) {
  return apiFetch<Listing>(`/listings/${id}/mark-available`, { method: "POST" });
}

/** Undoes mark-sold or pause. */
export function relistListing(id: string) {
  return apiFetch<Listing>(`/listings/${id}/relist`, { method: "POST" });
}

/** Reversible hide — undone by relistListing. */
export function pauseListing(id: string) {
  return apiFetch<Listing>(`/listings/${id}/pause`, { method: "POST" });
}

/** `listingIds` must be every listing in that shop — the API rejects a
 *  partial list rather than silently leaving gaps. */
export function reorderListings(shopId: string, listingIds: string[]) {
  return apiFetch<Listing[]>("/listings/reorder", {
    method: "POST",
    body: JSON.stringify({ shop_id: shopId, listing_ids: listingIds }),
  });
}

export function requestRestock(id: string) {
  return apiFetch<Listing>(`/listings/${id}/restock-request`, { method: "POST" });
}

export function withdrawRestockRequest(id: string) {
  return apiFetch<void>(`/listings/${id}/restock-request`, { method: "DELETE" });
}

/** Admin-only. `days: null` clears the promotion. */
export function featureListing(id: string, days: number | null) {
  return apiFetch<Listing>(`/listings/${id}/feature`, {
    method: "POST",
    body: JSON.stringify({ days }),
  });
}

export function uploadListingImage(id: string, file: File) {
  const formData = new FormData();
  formData.append("file", file);
  return apiFetch<ListingImage>(`/listings/${id}/images`, { method: "POST", body: formData });
}

/** Sets the image order. `imageIds` must be every image on the listing —
 *  the API rejects a partial list rather than silently leaving gaps. */
export function reorderListingImages(id: string, imageIds: string[]) {
  return apiFetch<Listing>(`/listings/${id}/images/reorder`, {
    method: "POST",
    body: JSON.stringify({ image_ids: imageIds }),
  });
}

export function deleteListingImage(id: string, imageId: string) {
  return apiFetch<void>(`/listings/${id}/images/${imageId}`, { method: "DELETE" });
}
